lib/Doctrine/DBAL/Driver/OCI8/OCI8Statement.php

```diff
@@ -114,27 +114,68 @@ public function __construct($dbh, $statement, OCI8Connection $conn)
      * @param string $statement The SQL statement to convert.
      *
      * @return string
+     * @throws \Doctrine\DBAL\Driver\OCI8\OCI8Exception
      */
     static public function convertPositionalToNamedPlaceholders($statement)
     {
-        $count = 1;
-        $inLiteral = false; // a valid query never starts with quotes
-        $stmtLen = strlen($statement);
         $paramMap = array();
-        for ($i = 0; $i < $stmtLen; $i++) {
-            if ($statement[$i] == '?' && !$inLiteral) {
-                // real positional parameter detected
-                $paramMap[$count] = ":param$count";
-                $len = strlen($paramMap[$count]);
-                $statement = substr_replace($statement, ":param$count", $i, 1);
-                $i += $len-1; // jump ahead
-                $stmtLen = strlen($statement); // adjust statement length
-                ++$count;
-            } elseif ($statement[$i] == "'" || $statement[$i] == '"') {
-                $inLiteral = ! $inLiteral; // switch state!
+        $count = 1;
+        $start = $offset = 0;
+        $fragments = array();
+        $quote = false;
+        $capture = function ($regex, $callback) use ($statement, $start, &$offset) {
+            if (preg_match($regex, $statement, $matches, PREG_OFFSET_CAPTURE, $offset)) {
+                $offset = $matches[0][1];
+                $callback($matches[0][0]);
+                return true;
+            }
+            return false;
+        };
+
+        do {
+            if (!$quote) {
```

Comment:
It can be a query which contains mixed parametrized and inline values coming from user input. I agree it's unlikely but not possible, especially in legacy applications.